### ISSUE TITLE
Ignore marker in directories

### ISSUE THREAD

**Issue by florianhofhammer** [syncthing/syncthing#8982]
Hi,

First of all, thanks for your great software!

I've run into issues when synchronizing directories for projects between multiple machines and changing build output directories for my synchronized source code directories frequently. For context, I typically do not want to synchronize build output; I'd rather rebuild on the corresponding machine if necessary.
Frequently changing the build directory name requires me to add the corresponding directory to my `.stignore` every time I change the directory, which can be quite cumbersome (and I might easily make typos there, forget a directory,...).

A feature that I'd like to propose is the exclusion of directories from synchronization based on the existence of a file (tagging that directory).
[Borg backup uses a similar feature](https://borgbackup.readthedocs.io/en/stable/usage/create.html?highlight=exclude-if-present) where directories are excluded from the backup if a specified file exists in the directory.
Such a feature would allow me (and any other user, of course) to just issue a quick `touch <exclude-tag-file>` in a directory and have Syncthing pick that up automatically (either via filesystem watchers or on the next filesystem scan). From then on, the directory would be excluded from synchronization until the marker is removed again.

Given that I'm not familiar with the Syncthing internals, I of course do not know how much effort it would be to add such a feature. Nevertheless, thanks in advance for considering this proposal!

**Comment by calmh** [comment]
This has been discussed previously but I can't find it so not calling it a duplicate yet... I agree this could be useful. I think it's not entirely trivial, we'd need to find and track all such makers and add their effect to the folder ignore patterns. But it's not impossible.

**Comment by fine-wind** [comment]
Hey, I know resilio sync is optional,
but Resilio Sync uses an undocumented, closed protocol with unknown security properties.
[see](https://docs.syncthing.net/users/faq.html#how-does-syncthing-differ-from-bittorrent-resilio-sync).
So, Syncthing add something similar?
The absence of publicly available information is always a concern.

**Comment by bolet** [comment]
I just asked for that feature [in the forum](https://forum.syncthing.net/t/add-rule-to-stignore-ignore-folders-containing-a-specific-file), now I discover it discussed here. Same reasons as @florianhofhammer and I'm also a borgbackup user.

**Comment by daviddavo** [comment]
That would also make it possible to ignore it depending on the device. For example, a `.mobileIgnore` file in a big folder would make that folder not sync to the laptop and smartphone, but sync between workstations.

**Comment by calmh** [comment]
If you mean what I think you mean, that highlights an interesting complication that probably wouldn't reliably work like you expect. (Ignoring a directory based on a marker in it, but the marker in it comes from the remote side, so needs to be synced, which requires the directory to not be ignored to begin with, and then what do we do when we've synced a bunch of files and discover one of them was a marker etc)

**Comment by cravl-dev** [comment]
One way to implement this I think would be pretty clean would be to allow stignore files to reference a parent directory, so you can automatically ignore any directory containing the file or folder marker. What's neat is that the user decides what to name such a file, and it could be different per-device just like any other ignore rule[1].

Syntactically, I can think of a couple ways you could implement this:

1. Use `..` to traverse up (example: `**/.nosync/../..`)
	- **Good:** Immediately familiar to most
	- **Bad:**
		- Requires special handling in code as a marker file will otherwise be treated as a dir and throw an error
		- Needing `/../..` to go one level up is a bit of a brain bender at first
		- Users might try to put `..` in the middle of a pattern
			- If allowed, introduces all kinds of issues with symlink resolution
			- If not allowed, is potentially confusing
2. Implement a specific structure (example: `**/(?=*/.nosync)`)
	- _This is emulating a regex lookahead; something like the CSS `:has` selector also comes to mind._
	- **Good:** More clear that it only works at the end
	- **Bad:** Maybe a bit ugly? Needs to be simple as possible without making existing stignores suddenly need escaping

This also handily covers a feature that I've been meaning to create a request for: the ability to ignore folders that are git-controlled. Pop in a `**/(?=*/.git)` and you're done!

[1] I only have a loose grasp of the internals, but I think it would work if all synced stignores are pre-synced before anything else, so that all peers know what the marker files are and can search for them in the file list before data is transferred, same as any other ignore pattern. EDIT: #10752 covers the needed mechanism, which exists in the API already. Also related: issue #9352, pull #10669.

**Comment by daviddavo** [comment]
Oh, right, you need the folder AND file to know that it should be ingored... Best case you will end up with an "empty" folder...

**Comment by acolomb** [comment]
I would like a solution as easy as `.git/..`(no need for any asterisk syntax AFAICT, didn't grasp what benefit your complicated special syntax options meant to provide).

But to set the expectations straight, this would be a whole complexity dimension added to the ignore handling code: So far, the premise is that a decision can be made from this info alone:

    ignore = F(path, patterns...)
        = F(path, pattern1) OR F(path, pattern2) OR F(path, pattern3) OR...

(With a short-circuit OR operator, and a bit simplified by neglecting negated patterns).

For any given path (simple text string), looking at each pattern in turn is sufficient to reach a final decision (first match wins). 

Any more elaborate rules as proposed here require a fundamentally different approach:

    ignore = F(path, patterns..., file_tree...)

If the path matches a pattern (e.g. `/some/dir/.git/..` matches `.git/..`), then we need to go back to `/some/dir` and mark it as ignored, while it previously was not because there was no matching pattern for it under this textual path. We can easily reach a contradiction if there is a negated pattern that also happens to apply for the same directory. It's no longer as easy as "first match wins", because each path may be affected by multiple patterns again (by aliasing it under different strings) *after* a decision was already reached for it (considering all patterns). That's where the rest of the file tree comes into the decision, when it previously wasn't needed.

And of course the next question is, "which file tree?" The local one during scanning? The recorded local state from previous scan? The tree as (partially) announced from a remote device whose index message we're currently trying to consolidate into the "global state"? Any combinations among these? A simple check for "does this path exist?" becomes a complex issue in itself. This part is the complex inversion of the easy question "does this path textually match that pattern (including glob rules)?"

At the very least, we would need two or more passes over the file tree, with a bit of caching for partial decisions. No way to answer immediately by looking at one path exactly once.

I'm not saying it's impossible, but it requires a vastly different approach to ignore processing at different points in the syncing process. Therefore it's quite unrealistic this will get tacked on as an extension to the current syntax system.

**Comment by calmh** [comment]
> And of course the next question is, "which file tree?" The local one during scanning? The recorded local state from previous scan? The tree as (partially) announced from a remote device whose index message we're currently trying to consolidate into the "global state"?

This one is especially unintuitive because no matter what you do it becomes weird.

- Let's say the tree of record is the one we have on disk. We might then sync a full directory from a remote device only to then discover that we've synced an ignore marker, and then suddenly ignore and orphan the directory. Undesirable.
- Let's say the tree of record is the global (announced) one as it exists in the database. We might then have a directory with data in it, but no ignore marker, but some other device has an ignore marker in it and we suddenly stop syncing it for no visible reason. Undesirable.

In short, this feature is simple and straight-forward when you only consider outgoing single-direction syncing, like Borg backup, rsync, etc, but nontrivial to even conceptualise for Syncthing...
